backend/msgpack/queue.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	errInstanceQueueClosed = errors.New("instance queue is closed")
+)
+
+// instanceQueue processes write requests for a given instance.
+type instanceQueue interface {
+	// Enqueue enqueues a write request.
+	Enqueue(writeOp *writeOp) error
+
+	// Close closes the queue.
+	Close() error
+}
+
+type writeFn func([]byte) error
+
+type queue struct {
+	sync.RWMutex
+
+	instance instance
+	log      xlog.Logger
+	conn     *connection
+	opsCh    chan *writeOp
+	closed   bool
+
+	writeFn writeFn
+}
+
+func newInstanceQueue(instance instance, opts ServerOptions) instanceQueue {
+	conn := newConnection(instance.Address(), opts.ConnectionOptions())
+	q := &queue{
+		instance: instance,
+		log:      opts.InstrumentOptions().Logger(),
+		conn:     conn,
+		opsCh:    make(chan *writeOp, opts.InstanceQueueSize()),
+	}
+	q.writeFn = q.conn.Write
+
+	go q.drain()
+	return q
+}
+
+func (q *queue) Enqueue(writeOp *writeOp) error {
+	q.RLock()
+	if q.closed {
+		q.RUnlock()
+		return errInstanceQueueClosed
+	}
+	// NB(xichen): the op buffer already batches multiple metric data points
+	// to maximize per packet utilization so there should be no need to perform
+	// additional batching here.
+	q.opsCh <- writeOp
```

Comment:
Channels are thread safe no? Does this have to be inside a RUnlock? Why not do a select write to channel - it will take care of the closed case as well as having the collector block on the snapshot call when this queue is full.